const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,279 @@
+of `impl const` items.
+
+## Implementation instructions
+
+1. Add an `is_const` field to the AST's `TraitRef`
+2. Adjust the Parser to support `const` modifiers before trait bounds
+3. Add an `is_const` field to the HIR's `TraitRef`
+4. Adjust lowering to pass through the `is_const` field from AST to HIR
+5. Add a a check to `librustc_typeck/check/wfcheck.rs` ensuring that all generic bounds
+    in an `impl const` block have the `in_const` flag set and all methods' `constness` field is
+    `Const`.
+6. Feature gate instead of ban `Predicate::Trait` other than `Sized` in
+    `librustc_mir/transform/qualify_min_const_fn.rs`
+7. Remove the call in https://github.com/rust-lang/rust/blob/f8caa321c7c7214a6c5415e4b3694e65b4ff73a7/src/librustc_passes/ast_validation.rs#L306
+8. Adjust the reference and the book to reflect these changes.
+
+## Const type theory
+
+This RFC was written after weighing practical issues against each other and finding the sweet spot
+that supports most use cases, is sound and fairly intuitive to use. A different approach from a
+type theoretical perspective started out with a much purer scheme, but, when exposed to the
+constraints required, evolved to essentially the same scheme as this RFC. We thus feel confident
+that this RFC is the minimal viable scheme for having bounds on generic parameters of const
+functions. The discussion and evolution of the type theoretical scheme can be found
+[here](https://github.com/rust-rfcs/const-eval/pull/8#issuecomment-452396020) and is only 12 posts
+and a linked three page document long. It is left as an exercise to the reader to read the
+discussion themselves.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It is not a fully general design that supports every possible use case,
+but it covers the most common cases. See also the alternatives.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Effect system
+
+A fully powered effect system can allow us to do fine grained constness propagation
+(or no propagation where undesirable). This is way out of scope in the near future
+and this RFC is forward compatible to have its background impl be an effect system.
+
+## Fine grained `const` annotations
+
+One could annotate methods instead of impls, allowing just marking some method impls
+as const fn. This would require some sort of "const bounds" in generic functions that
+can be applied to specific methods. E.g. `where <T as Add>::add: const` or something of
+the sort. This design is more complex than the current one and we'd probably want the
+current one as sugar anyway
+
+## No explicit `const` bounds
+
+One could require no `const` on the bounds (e.g. `T: Trait`) and assume constness for all
+bounds. An opt-out via `T: ?const Trait` would then allow declaring bounds that cannot be
+used for calling methods. This design causes discrepancies with `const fn` pointers as
```

Comment:
The discrepancy is that 

```rust
impl const Trait for Type {
    fn foo<T: OtherTrait>(t: T) { ... }
}
```

would require `OtherTrait` to be `const` automatically, while

```rust
const fn foo<T: OtherTrait>(t: T) { ... }
```

would not do that. That is very inconsistent and will lead to a lot of confusion.

For semver purposes (which is the only reason we even have the `const` keyword for function and are considering it for bounds and impls) we either need all bounds on methods of an `impl const Trait` to be `const`, too, or we need none of them to be `const`, which would be very sad because it would make `impl const Hash` impossible to implement (other than outright panicking or making it do nothing, which is wrong).

Reply:
Trait bounds on `const fn`s aren't stabilized yet, so doesn't this just mean that we should apply the same rules there, and have `const fn foo<T: OtherTrait>(t: T)` require `OtherTrait` to be `const` unless called at runtime?

I am not sure what the last part of your comment means. `impl const Hash for T` does need `hash`'s `H: Hasher` to be `const`, sure, but that doesn't mean *every* trait will *always* be like that, or that *every* trait will *always* want this "`const` polymorphism" to remove that requirement at runtime.

Reply:
> every trait will always be like that, or that every trait will always want this "const polymorphism" to remove that requirement at runtime.

Yes, I see the problems with my approach now. Thanks